core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -787,8 +787,37 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
 
   final def left[R1 <: R, C]: ZIO[Either[R1, C], E, Either[A, C]] = self +++ ZIO.identity[C]
 
+  /**
+   * Returns a successful effect if the value is 'Left', or fails with the error f.
+   */
+  def leftOrFail[B, C, E1 >: E](f: E1)(implicit ev: A <:< Either[B, C]): ZIO[R, E1, B] =
+    self.flatMap(ev(_).fold[ZIO[R, E1, B]](ZIO.succeed(_), _ => ZIO.fail(f)))
+
+  /**
+   * Returns a successful effect if the value is 'Left', or fails with a [[NoSuchElementException]].
+   */
+  def leftOrFailException[B, C, E1 >: NoSuchElementException](
```

Comment:
Hmm why `E1 >: NoSuchElementException` rather than just returning `ZIO[R, NoSuchElementException, B]`? Since `E1` doesn't show up in the function's parameters (other than the `ev2` but that doesn't influence inference, I think), this might not infer properly (or worse inferred as `Any`)